### ISSUE TITLE
Can't query $cmd collection

### ISSUE THREAD

**Issue by alcaeus** [mongodb/mongo-php-driver#329]
This was previously reported in alcaeus/mongo-php-adapter#110, from the trace I'd assume the error originates in the driver.

MongoDB exposes a virtual `$cmd` collection that can be used to run database commands. So, the following query is identical to the `currentOp` command:
```
> db.$cmd.sys.inprog.findOne()
{
	"inprog" : [
		{
			"desc" : "conn5",
			"threadId" : "0x7000008b1000",
			"connectionId" : 5,
			"client" : "127.0.0.1:55488",
			"active" : true,
			"opid" : 18107,
			"secs_running" : 0,
			"microsecs_running" : NumberLong(24),
			"op" : "command",
			"ns" : "admin.$cmd",
			"query" : {
				"currentOp" : 1
			},
			"numYields" : 0,
			"locks" : {
				
			},
			"waitingForLock" : false,
			"lockStats" : {
				
			}
		}
	],
	"ok" : 1
}
```

However, running this via mongo-php-library will cause a driver exception:
```
php > $client = new MongoDB\Client();
php > $client->selectCollection('admin', '$cmd.sys.inprog')->findOne();
PHP Warning:  Uncaught exception 'MongoDB\Driver\Exception\RuntimeException' with message 'Invalid collection name: admin.$cmd.sys.inprog' in /Users/alcaeus/Code/mongo-php-library/src/Operation/Find.php:179
Stack trace:
#0 /Users/alcaeus/Code/mongo-php-library/src/Operation/Find.php(179): MongoDB\Driver\Server->executeQuery('admin.$cmd.sys....', Object(MongoDB\Driver\Query), Object(MongoDB\Driver\ReadPreference))
#1 /Users/alcaeus/Code/mongo-php-library/src/Operation/FindOne.php(81): MongoDB\Operation\Find->execute(Object(MongoDB\Driver\Server))
#2 /Users/alcaeus/Code/mongo-php-library/src/Collection.php(468): MongoDB\Operation\FindOne->execute(Object(MongoDB\Driver\Server))
#3 php shell code(1): MongoDB\Collection->findOne()
#4 {main}
  thrown in /Users/alcaeus/Code/mongo-php-library/src/Operation/Find.php on line 179

Warning: Uncaught exception 'MongoDB\Driver\Exception\RuntimeException' with message 'Invalid collection name: admin.$cmd.sys.inprog' in /Users/alcaeus/Code/mongo-php-library/src/Operation/Find.php:179
Stack trace:
#0 /Users/alcaeus/Code/mongo-php-library/src/Operation/Find.php(179): MongoDB\Driver\Server->executeQuery('admin.$cmd.sys....', Object(MongoDB\Driver\Query), Object(MongoDB\Driver\ReadPreference))
#1 /Users/alcaeus/Code/mongo-php-library/src/Operation/FindOne.php(81): MongoDB\Operation\Find->execute(Object(MongoDB\Driver\Server))
#2 /Users/alcaeus/Code/mongo-php-library/src/Collection.php(468): MongoDB\Operation\FindOne->execute(Object(MongoDB\Driver\Server))
#3 php shell code(1): MongoDB\Collection->findOne()
#4 {main}
  thrown in /Users/alcaeus/Code/mongo-php-library/src/Operation/Find.php on line 179
```

**Comment by jmikola** [comment]
Since this is a RuntimeException, I believe it's originating from [the server](https://github.com/mongodb/mongo/search?l=cpp&q=%22Invalid+collection+name%22&type=Code&utf8=%E2%9C%93). Can you confirm the server version that this occurs on? I assume it's 3.2+, where the new find command would reject such a collection name.

You'll want to use the `currentOp` command on MongoDB 3.2 (see the shell definition of [`db.currentOp`](https://docs.mongodb.com/manual/reference/method/db.currentOp/) for details). While we do document the shell method, the command itself is not documented. I've opened [DOCS-8007](https://jira.mongodb.org/browse/DOCS-8007) to track that.

Additionally, I tested the same query on MongoDB 3.0.x and found that it results in a server-side assertion, which immediately drops the connection and results in a read timeout on the client. This is due to the PHPC driver wrapping its query filters in `$query` unnecessarily. I've opened [PHPC-705](https://jira.mongodb.org/browse/PHPC-705) to track that. Upstream, I've opened [SERVER-24341](https://jira.mongodb.org/browse/SERVER-24341) to discuss the server's assertion. They previously fixed a similar issue with asserting on the presence of `$readPreference`, which should have been ignored.

Note: as-is, there is no way to query for the current op from PHPC.

----

FWIW, libmongoc uses `mongoc_collection_command()` to issue legacy current op queries, as demonstrated in https://github.com/mongodb/mongo-c-driver/commit/ca623c1491b2b16d5648b70d4cbcfa6b3fd84656 (expand the diff below the changed lines); however, that was later removed from the test suite in https://github.com/mongodb/mongo-c-driver/commit/1ad5c28458a39238d313d9b41796f21ed9193a43 (for [CDRIVER-888](https://jira.mongodb.org/browse/CDRIVER-888)). The method is still present on the collection API, but we don't utilize it from PHPC.

**Comment by alcaeus** [comment]
> Can you confirm the server version that this occurs on? I assume it's 3.2+, where the new find command would reject such a collection name.

Correct, the results above were produced on MongoDB 3.2.6. Running the command directly via the API instead of querying the `$cmd` collection was my suggestion as well.

Thank you for creating the bugs upstream 👍

**Comment by jmikola** [comment]
Note: this may be addressed by [PHPC-783](https://jira.mongodb.org/browse/PHPC-783). I've linked the PHPC JIRA tickets so we can confirm later.

**Comment by jmikola** [comment]
@alcaeus: Confirming that this is fixed in master and will be in the next 1.2.0-dev release.

See: [PHPC-705](https://jira.mongodb.org/browse/PHPC-705) and the regression tests in https://github.com/mongodb/mongo-php-driver/pull/420.
